gloo/transport/ibverbs/unbound_buffer.cc

```diff
@@ -0,0 +1,257 @@
+/**
+ * Copyright (c) 2018-present, Facebook, Inc.
+ * All rights reserved.
+ *
+ * This source code is licensed under the BSD-style license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+
+#include "gloo/transport/ibverbs/unbound_buffer.h"
+
+#include <cstring>
+
+#include "gloo/common/error.h"
+#include "gloo/common/logging.h"
+#include "gloo/transport/ibverbs/context.h"
+
+namespace gloo {
+namespace transport {
+namespace ibverbs {
+
+UnboundBuffer::UnboundBuffer(
+    const std::shared_ptr<Context>& context,
+    void* ptr,
+    size_t size)
+    : ::gloo::transport::UnboundBuffer(ptr, size),
+      context_(context),
+      recvRank_(-1),
+      sendCompletions_(0),
+      sendRank_(-1),
+      shareableNonOwningPtr_(this) {
+  std::unique_lock<std::mutex> lock(m_);
+  auto dev = context->device_;
+
+  auto mr = ibv_reg_mr(
+      dev->pd_,
+      // Empty buffers still need a valid pointer and positive length otherwise
+      // IB throws an error. We use our size 1 empty buffer for this.
+      size == 0 ? static_cast<void*>(&emptyBuf_) : ptr,
+      size == 0 ? sizeof(emptyBuf_) : size,
+      IBV_ACCESS_LOCAL_WRITE | IBV_ACCESS_REMOTE_READ |
+          IBV_ACCESS_REMOTE_WRITE);
+
+  // Provide hint if the error is EFAULT and nv_peer_mem is not loaded
+  if (mr == nullptr && errno == EFAULT) {
+    if (!dev->hasNvPeerMem_) {
+      GLOO_ENFORCE(
+          mr != nullptr,
+          "ibv_reg_mr: ",
+          strerror(errno),
+          " (kernel module 'nv_peer_mem' not loaded;"
+          " did you specify a pointer to GPU memory?)");
+    }
+  }
+
+  // Provide hint if the error is ENOMEM
+  if (mr == nullptr && errno == ENOMEM) {
+    GLOO_ENFORCE(
+        mr != nullptr,
+        "ibv_reg_mr: ",
+        strerror(errno),
+        " (did you run into the locked memory limit?)");
+  }
+
+  GLOO_ENFORCE(
+      mr != nullptr,
+      "ibv_reg_mr: ",
+      strerror(errno),
+      " ptr=",
+      ptr,
+      " size=",
+      size);
+
+  mr_ = mr;
+}
+
+UnboundBuffer::~UnboundBuffer() {
+  std::lock_guard<std::mutex> guard(m_);
+
+  ibv_dereg_mr(mr_);
+}
+
+void UnboundBuffer::abortWaitRecv() {
+  std::lock_guard<std::mutex> guard(m_);
+  abortWaitRecv_ = true;
+  recvCv_.notify_one();
```

Comment:
IIUC, this does guarantee which thread to wakes up. Would this be a problem? Oh maybe it is per pair, so it won't be a problem?

Reply:
abortWaitRecv only aborts a single wait request so we only want to wake one thread. I don't think we ever call multiple waitRecv on the same buffer. Usually each algorithm is only waiting for one thing at a time since it's a blocking call